msdk-spectra/msdk-spectra-isotopepattern/src/main/java/io/github/msdk/spectra/isotopepattern/IsotopePatternGeneratorAlgorithm.java

```diff
@@ -151,32 +151,43 @@
      * intensity is added together and new m/z value is calculated as a weighted
      * average.
      */
-    private static int mergeIsotopes(double mzValues[], float intensityValues[],
-            int size, double mzTolerance) {
-
-        int newSize = 0;
-
-        for (int i = 0; i < size; i++) {
-
-            if ((i < size - 1) && (Math
-                    .abs(mzValues[i] - mzValues[i + 1]) < mzTolerance)) {
-                float newIntensity = intensityValues[i]
-                        + intensityValues[i + 1];
-                double newMZ = (mzValues[i] * intensityValues[i]
-                        + mzValues[i + 1] * intensityValues[i + 1])
-                        / newIntensity;
-                mzValues[newSize] = newMZ;
-                intensityValues[newSize] = newIntensity;
-                mzValues[i + 1] = newMZ;
-                intensityValues[i + 1] = newIntensity;
-                i++;
+    private static int mergeIsotopes(double mzValues[], float intensityValues[], int size, double mzTolerance) {
+        if (mzValues.length != intensityValues.length)
+            throw new IllegalArgumentException("Mass and intensity arrays must be of the same size");
+        if (mzValues.length < 2)
+            return mzValues.length;
+
+        int ptrCur = 0;
+        int ptrNex = 1;
+        for ( ; ptrNex < mzValues.length; ptrNex++) {
+            double mzCur = mzValues[ptrCur];
+            double mzNex = mzValues[ptrNex];
+            if (Math.abs(mzCur - mzNex) < mzTolerance) {
+                // merge, only next pointer moves
+                float abCur = intensityValues[ptrCur];
+                float abNex = intensityValues[ptrNex];
+                float abNew = abCur + abNex;
+                double mzNew = (mzCur * abCur + mzNex * abNex) / abNew;
+                mzValues[ptrCur] = mzNew;
+                intensityValues[ptrCur] = abNew;
+
             } else {
-                mzValues[newSize] = mzValues[i];
-                intensityValues[newSize] = intensityValues[i];
+                // don't merge, move current pointer and copy the value there
+                ptrCur++;
+
+                mzValues[ptrCur] = mzValues[ptrNex];
+                intensityValues[ptrCur] = intensityValues[ptrNex];
             }
-            newSize++;
         }
-        return newSize;
+
+        // ugly hack to make SimpleSpectrum not complain about masses not being sorted
```

Comment:
Since arrays are not easily resized in Java, we always use the trio of mzValues[], intensityValues[], and int size. The size indicates the real size of the array, while the length of the actual arrays might be different. Note that the SimpleMsSpectrum only checks the proper ordering of m/z values up to the given size, not the whole array:
https://github.com/msdk/msdk/blob/master/msdk-datamodel/src/main/java/io/github/msdk/datamodel/impl/SimpleMsSpectrum.java#L113
Also the DataPointSorter.sortDataPoints() sorts everything up to given size and ignores the rest.